### PR TITLE
The new classes IPv4Address, IPv6Address and IPAddress (discussion)

### PR DESCRIPTION
**Pull request by gx740**
I have developed the new lightweight classes for manipulating the IP addresses. These are very fast and use less memory.

I have made the tests for the most common use case: creation the address out of byte array (for example: IP/TCP reassembly logic). The tests ran under Core i7 9700K, Win10(x64):
```
=== Current ===
Sizeof IPv4 = 68
Sizeof IPv6 = 80
IPv4 (uint): duration(ms) = 1803, iterations = 10.000.000, calls per sec = 5.546.000
IPv6: duration(ms) = 4411, iterations = 10.000.000, calls per sec = 2.267.000

=== New ===
Sizeof IPv4 = 4
Sizeof IPv6 = 16
Sizeof IP = 24
IPv4 (uint): duration(ms) = 12, iterations = 10.000.000, calls per sec = 833.333.000
IPv6: duration(ms) = 32, iterations = 10.000.000, calls per sec = 312.500.000
```

Replacing the old classes to the new ones will take some time so I placed the new classes into namespace `experimental`. First of all I am planning to change the `IPReassembly` and `TCPReassembly`.

Is any chance that this PR will be merged?

### CODE REVIEW

**Comment by gx740** [comment]
Source code of the tests:

```c++
PTF_TEST_CASE(TestIPAddrsPerfomance)
{
  constexpr auto tryCount = 10000000U;
  uint32_t const uintIPv4 = htonl(0x0A000004);
  uint8_t const bytesIPv6[16] = { 0x26, 0x07, 0xF0, 0xD0, 0x10, 0x02, 0x00, 0x51, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04 };

  // current implementation
  {
    std::vector<IPv4Address> ipv4Addrs;
    std::vector<IPv6Address> ipv6Addrs;
    ipv4Addrs.reserve(tryCount);
    ipv6Addrs.reserve(tryCount);

    std::cout << "=== Current ===\n";

    std::cout << "Sizeof IPv4 = " << sizeof(IPv4Address) + 4 << '\n';
    std::cout << "Sizeof IPv6 = " << sizeof(IPv6Address) + 16 << '\n';

    // ipv4
    uint64_t count = 0;
    auto startTime = std::chrono::steady_clock::now();

    for(auto i = 0U; i < tryCount; ++i, ++count)
      ipv4Addrs.emplace_back(uintIPv4);

    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - startTime);
    std::cout << "IPv4 (uint): duration(ms) = " << duration.count()
      << ", iterations = " << count
      << ", calls per sec = " << (count / duration.count()) * 1000
      << '\n';
    //std::cout << "Addr4 = " << ipv4Addrs[100].toString() << '\n';

    // ipv6
    count = 0;
    startTime = std::chrono::steady_clock::now();
    
    for(auto i = 0U; i < tryCount; ++i, ++count)
      ipv6Addrs.emplace_back(const_cast<uint8_t *>(bytesIPv6));

    duration = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - startTime);
    std::cout << "IPv6: duration(ms) = " << duration.count()
      << ", iterations = " << count
      << ", calls per sec = " << (count / duration.count()) * 1000
      << '\n';
    //std::cout << "Addr6 = " << ipv6Addrs[100].toString() << '\n';
  }


  // new versions
  {
    std::vector<pcpp::experimental::IPv4Address> ipv4Addrs;
    std::vector<pcpp::experimental::IPv6Address> ipv6Addrs;
    ipv4Addrs.reserve(tryCount);
    ipv6Addrs.reserve(tryCount);

    std::cout << "=== New ===\n";
    std::cout << "Sizeof IPv4 = " << sizeof(pcpp::experimental::IPv4Address) << '\n';
    std::cout << "Sizeof IPv6 = " << sizeof(pcpp::experimental::IPv6Address) << '\n';
    std::cout << "Sizeof IP = " << sizeof(pcpp::experimental::IPAddress) << '\n';

    // ipv4
    uint64_t count = 0;
    auto startTime = std::chrono::steady_clock::now();

    for(auto i = 0U; i < tryCount; ++i, ++count)
      ipv4Addrs.emplace_back(uintIPv4);

    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - startTime);
    std::cout << "IPv4 (uint): duration(ms) = " << duration.count()
      << ", iterations = " << count
      << ", calls per sec = " << (count / duration.count()) * 1000
      << '\n';
    //std::cout << "Addr4 = " << ipv4Addrs[100].toString() << '\n';

    // ipv6
    count = 0;
    startTime = std::chrono::steady_clock::now();

    for(auto i = 0U; i < tryCount; ++i, ++count)
      ipv6Addrs.emplace_back(const_cast<uint8_t *>(bytesIPv6));

    duration = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - startTime);
    std::cout << "IPv6: duration(ms) = " << duration.count()
      << ", iterations = " << count
      << ", calls per sec = " << (count / duration.count()) * 1000
      << '\n';
    //std::cout << "Addr6 = " << ipv6Addrs[100].toString() << '\n';
  }
  std::cout << "=== End ===\n";
}
```

**Comment by gx740** [comment]
I think, in order to migrate to the new classes the temporary branch should be created. After the changes will be made this branch must be incorporated into `dev`.
Notice I have prepared the changes in `IPReassembly` and `TCPReassembly`.

**Comment by seladb** [comment]
Let me take a look and get back to you. Ideally we shouldn't have multiple IPv4/IPv6 classes because it'll be confusing to the users. If we can preserve the same (or at least almost the same) API that'd be good.

I'll take a look and let you know what I think

**Comment by gx740** [comment]
Yes, only one implementation should stay in project

**Comment by seladb** [comment]
I went briefly over the new implementation, it looks good!

If I understand correctly the main change is around the inheritance order, so instead of having `IPv4Address` and `IPv6Address` inherit from `IPAddress` you now have them as distinct classes and have `IPAddress` as kind of a facade. Is that correct?

**Comment by gx740** [comment]
Yes, it's correct.

Notice the implementation of the classes is based on document ISO N4771 "Working Draft, C++ Extensions for Networking".

**Comment by seladb** [comment]
I'm not familiar with this document. Has this proposal been implemented in C++ STL?

My suggestion would be to try to have the new implementation replacing the old one. Most of the usages in PcapPlusPlus refer to either `IPv4Address` specifically, `IPv6Address` specifically or a pointer/reference to `IPAddress`. It seems to me that replacing the underlying implementation should be a very big deal. What do you think?

**Comment by gx740** [comment]
> I'm not familiar with this document. Has this proposal been implemented in C++ STL?

It's not implemented yet but I think it will be done in the future. Anyway, this is a good implementation.

I am ready to make the changes in all modules. We should decide how to do that:
1. We can make a temporary branch and make all the changes in it and then merge them into `dev` or
2. Place the new classes into the new namespace (for example: `common`) and make the changes in branch `dev`

**Comment by seladb** [comment]
How big do you think the change is? Ideally if you keep the same API, at least for `IPv4Address` and `IPv6Address` the change should be very minimal right?

**Comment by gx740** [comment]
I can append the methods from the old classes to save backward compatibility but we must mark these as deprecated. Later we can remove them.

**Comment by gx740** [comment]
I have prepared the changes in the following modules (including the depending unit tests and examples):
- TCP reassembly
- IP reassembly
- ICMP layer
- IGMP layer
- ARP layer
- NetworkUtils

**Comment by seladb** [comment]
It doesn't seem to be a huge amount of work, but at the same time it doesn't seem small. Ideally we should keep the interface as backward compatible as possible and make changes only where it's necessary.

From your analysis, where do you see most of the changes? Are they in modules that use `IPv4Address`/`IPv6Address` or are they in modules that use `IPAddress`?

**Comment by gx740** [comment]
The modules that depend from the certain class:

**IPAddress**

- PcapRemoteDevice
- PcapRemoteDeviceList
- RawSocketDevice
- PcapFilter
- PcapLiveDeviceList
- TCPReassembly

**IPv4Address**

- DHCPLayer
- ICMPLayer
- IGMPLayer
- IPReassmbly
- TCPReassembly
- RadiusLayer
- SDPLayer
- ARPLayer
- DNSResourceData
- IPv4Layer
- TLVData
- NetworkUtils
- PcapLiveDevice
- PcapLiveDeviceList
- PcapFilter
- PcapRemoteDeviceLIst

**IPv6Address**

- IPReassmbly
- TCPReassembly
- IPv6Extensions
- IPv6Layer
- DNSResourceData
- PcapRemoteDeviceLIst
- PcapLiveDeviceList
- PcapFilter

I have prepared the changes to keep backward compatibility as possible.

In fact, the changes that should be made are simple for understanding.

**Comment by seladb** [comment]
You mean that all of those files have to be changed?

**Comment by gx740** [comment]
Yes.

**Comment by seladb** [comment]
That's a pretty big change... is there any way to minimize the change?

**Comment by gx740** [comment]
No, there is no way.

The most of changes are very simple like:
- replace one name to another
- change the method name
- remove unnecessary operations

It's not a problem at all besides I will do all work.
Yes, it will take some time but we can migrate gradually in a new branch and, when all changes have completed, incorporate into `dev`.
Different way is to place classes into namespace and make the changes in `dev`. In this case, the changes will available to the users at once.

**Comment by gx740** [comment]
The changes have been prepared for all modules except the following:
- PcapFilter
- PcapLiveDevice
- PcapLiveDeviceList
- PcapRemoteDeviceList
- PcapRemoteDevice

**Comment by seladb** [comment]
Ok that makes sense. You can do the work in another branch and later we'll merge it into `dev`. 

To make it easier to review, it'd be great if you can build it in chunks, meaning: build some of it and let me review, and then move on to build more and let me review, and so on.

Please let me know what you think

**Comment by gx740** [comment]
Currently, I have made a number of branches in which the changes are grouped by certain module/logic.

I think you should create a branch based on `dev` and all changes I will compare with it. At the end this branch should be incorporated into `dev`. Is any problem with this approach?

**Comment by seladb** [comment]
Sure, I agree. I've just created `ipaddr_refactor` out of `dev`. We can merge the PRs to that branch

File: Packet++/src/EthDot3Layer.cpp [hunk 1]
```diff
@@ -1,3 +1,9 @@
+#if defined(WIN32) || defined(WINx64) || defined(PCAPPP_MINGW_ENV)
+#include <winsock2.h>
+#else
+#include <netinet/in.h>
+#endif
+
```
Comment (seladb) [inline Packet++/src/EthDot3Layer.cpp]: You can remove that, I've just pushed a fix that handle those issues

**Comment by gx740** [comment]
Cancel the checks for this PR please.

**Comment by seladb** [comment]
I'm not sure how to cancel the checks for a specific PR. One thing that you can do is create a branch that forks from `ipaddr_refactor` and disable CI there by changing the CI config file names (`.travis.yml`, `appveyor.yml`, `.cirrus.yml`)

**Pull request closed without merging**